File: bindings/node/perf/multi/perf_multi_dealer_router_server.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const readline = require('node:readline');
const zlink = require('../../..');
const { parseMultiArgs } = require('./perf_multi_common');
const {
  POLLIN,
  POLLOUT,
  applyAutoHwmMsgUnit,
  applyContextPolicy,
  applySocketPolicy,
  pollEvents,
  pollEventHas,
  recvNoWait,
  resolveClientPollTimeoutMs,
  trySocketSend,
  waitForConnectionReadyCount
} = require('./perf_multi_runtime');

async function main() {
  const options = parseMultiArgs(process.argv.slice(2));
  const ctx = new zlink.Context();
  applyContextPolicy(ctx, 'server', 'MULTI_DEALER_ROUTER');
  const router = new zlink.RouterSocket(ctx);
  const poller = new zlink.Poller();
  const pending = [];
  let rl = null;
  let stop = false;

  try {
    applySocketPolicy(router);
    router.bind(options.endpoint);
    applyAutoHwmMsgUnit(router, options.msgSize);
    ctx.recalculateAutoHwm();
    poller.add(router, pollEvents(POLLIN));
    const readyBarrier = waitForConnectionReadyCount(router, options.clients);
    console.log(`READY,${options.endpoint}`);

    rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
    (async () => {
      for await (const line of rl) {
        if (line === 'STOP' || line === 'QUIT') {
          stop = true;
          break;
        }
      }
    })();

    await readyBarrier;
    const clientPollTimeoutMs = resolveClientPollTimeoutMs();

    while (!stop) {
      poller.modify(router, pollEvents(pending.length > 0 ? (POLLIN | POLLOUT) : POLLIN));
      const ready = poller.wait(clientPollTimeoutMs);
      if (!ready) {
        continue;
      }

      if (pollEventHas(ready, POLLIN)) {
        while (true) {
          const received = recvNoWait(router);
          if (!received) {
            break;
          }
          try {
            if (!received.routingId) {
              received.close();
              continue;
            }
            if (pending.length === 0 && trySocketSend(router, received.routingId, received.parts)) {
              received.close();
              continue;
            }
            pending.push(received);
          } catch (error) {
            received.close();
            throw error;
          }
        }
      }

      if (pollEventHas(ready, POLLOUT)) {
        while (pending.length > 0) {
          const current = pending[0];
          if (!trySocketSend(router, current.routingId, current.parts)) {
            break;
          }
          pending.shift();
          current.close();
        }
      }
    }
  } finally {
    rl?.close();
    while (pending.length > 0) {
      pending.shift().close();
    }
    poller.close();
    router.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: bindings/python/perf/multi/perf_multi_dealer_router_server.py
import sys
import threading
from collections import deque

import zlink

from perf_multi_common import (
    apply_multi_socket_options,
    benchmark_endpoint,
    parse_server_args,
    perf_server_context,
    recv_nonblocking,
    safe_poll,
    send_nonblocking,
)


def main(argv=None):
    args = parse_server_args(argv or sys.argv[1:])
    endpoint = benchmark_endpoint(args.transport, "multi-dealer-router")
    stop = threading.Event()
    pending = deque()

    def wait_stop():
        for line in sys.stdin:
            if line.strip().upper() in {"STOP", "QUIT"}:
                stop.set()
                return
        # stdin EOF (parent closed pipe) is also a STOP signal.
        stop.set()

    threading.Thread(target=wait_stop, daemon=True).start()

    with perf_server_context() as ctx:
        with zlink.RouterSocket(ctx) as router:
            apply_multi_socket_options(router)
            router.bind(endpoint)
            print(f"READY,{endpoint}", flush=True)
            with zlink.Poller() as poller:
                poller.add_socket(
                    router,
                    zlink.PollEventFlag.POLLIN | zlink.PollEventFlag.POLLOUT,
                )
                recv_storage = zlink.Received()
                # PERF_MULTI_TEST_POLICY § 1.3.1: signal-driven wait. The
                # echo server has no in-band phase end of its own; the
                # runner shuts it down via stdin STOP/EOF and SIGTERM.
                # We wake immediately on POLLIN/POLLOUT so we never sit on
                # a short timer cadence for I/O readiness.
                while not stop.is_set():
                    while pending:
                        routing_id, payload = pending[0]
                        if not send_nonblocking(router, payload, routing_id=routing_id):
                            break
                        pending.popleft()
                    while True:
                        received = recv_nonblocking(router, storage=recv_storage)
                        if received is None:
                            break
                        with received:
                            payload = received.to_bytes_list()[0]
                            routing_id = bytes(received.routing_id)
                            sent = False
                            if not pending:
                                sent = (
                                    received.send()
                                    .message(payload)
                                    .flags(zlink.SendFlags.DONT_WAIT)
                                    .submit()
                                )
                        if not sent:
                            pending.append((routing_id, payload))
                    safe_poll(poller, -1)


if __name__ == "__main__":
    main()
